### PR TITLE
clang-format 18

### PR DESCRIPTION
**Pull request by graebm**
**Issue:**
- We've used clang-format 9 since forever, but it's not easy for people to get such an old version anymore.
- You really need everyone on the same exact version clang-format, even PATCH matters
- It's very hard to get the same version on everyone's machine (Ubuntu 24 has 18, brew on MacOS has 18, but Amazon Linux 2 only has 11, and AL2023 only has 15)

**Description of Changes:**
- start using latest version of clang-format: 18.1.6
- use [pipx](https://github.com/pypa/pipx) to install and run [clang-format from PyPI](https://pypi.org/project/clang-format). This lets everyone run the same version, regardless of OS
- rewrite ~format-check.sh~ as `format-check.py`
    - now we can run it on Windows
    - most people on this team know Python better than Bash
    - add `-i` option to edit files in place
    - the script runs the specific clang-format version via pipx, so even if you have some other version of clang-format installed, the script will get it right
- CI just runs `./format-check.py` instead of using some 3rdparty Github Action

**Updating your Machine:**

- MacOS:
    - if you have clang-format 9 from homebrew, uninstall it: `brew uninstall llvm@9`
    - if you have another version from homebrew, uninstall it: `brew uninstall clang-format`
    - install pipx: `brew install pipx`
    - ensure programs that pipx install are on your $PATH: `pipx ensurepath`
    - install the correct clang-format, so your IDE can use it too: `pipx install clang-format==18.1.6`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by graebm** [comment]
Oh hold until new `ubuntu-24.04` Github runner becomes available...
https://github.com/actions/runner-images/issues/9691

File: include/aws/common/error.h [hunk 1]
```diff
@@ -36,12 +36,20 @@ struct aws_error_info_list {
     uint16_t count;
 };
 
+/* clang-format off */
+/* reenable formatting once we're all using 18.1.5+ */
+
```
[the remark this answers is not in the document]
Comment (waahm7) [reply include/aws/common/error.h:41]: Interesting: `You can use pipx to run clang-format, as well. For example, pipx run clang-format <args> will run clang-format without any previous install required on any machine with pipx (including all default GitHub Actions / Azure runners, avoiding requiring a pre-install step or even actions/setup-python).`

[the remark this answers is not in the document]
Comment (graebm) [reply include/aws/common/error.h:41]: I already fixed up all the other repos on my machine. There's just one code pattern that's causing all these differences between 18.1.3 and 18.1.5. It's when we use a macro to init the fields of a structure. aws-c-common has way more instances of this than any other repo

[the remark this answers is not in the document]
Comment (graebm) [reply include/aws/common/error.h:41]: I'll look into the `pip` thing...

[the remark this answers is not in the document]
Comment (graebm) [reply include/aws/common/error.h]: pipx is great, lets everyone use the same clang-format version! Thanks!

File: format-check.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python3
+import argparse
+import os
+from pathlib import Path
+import re
+from subprocess import list2cmdline, run
+from tempfile import NamedTemporaryFile
+
```
Comment (waahm7) [inline format-check.py:8]: trivial:
```suggestion

# 18.1.6 is the latest version as of June 2024
```

**Review by waahm7 — APPROVED** [review]
🥳 🎆🎆 🎊🎊🎊🥳

**Pull request merged**